KEP-4008: CRDValidationRatcheting: Add support for optional `oldSelf`

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

NONE

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adds `optionalOldSelf` to `x-kubernetes-validations` to support ratcheting CRD schema constraints
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/4008
```

/cc @cici37

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -221,7 +223,11 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+		// If ratcheting is enabled, allow rule with oldSelf to evaluate
+		// when `optionalOldSelf` is set to true
+		allowsNilOldSelf := utilfeature.DefaultFeatureGate.Enabled(features.CRDValidationRatcheting) && rule.OptionalOldSelf != nil && !*rule.OptionalOldSelf
+		if compiled.UsesOldSelf && oldObj == nil && !allowsNilOldSelf {
```
And you automatically get `oldSelf` as being nil in the following code?

alexzielenski replied:
No, had to update the activation factory to also return `true` for `hasOldSelf`

apelisse reviewed (commented):
I'd love to see a test.

alexzielenski commented:
tests added

jiahuif commented:
/assign
/assign @cici37 
/triage accepted

alexzielenski committed:
codegen

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,9 @@ type ValidationRule struct {
 	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+
+	// +optional
+	OptionalOldSelf *bool `json:"optionalOldSelf,omitempty"`
```
Is the field itself feature gated? The name is confusing. Since it is added into a stable API, it should follow the feature gate policy maybe?

alexzielenski replied:
The field only does anything if the feature gate is enabled. Is that what you mean? Or you mean it should throw a validation error if gate is not enabled?

cici37 replied:
I mean adding the field itself might also need to be feature gated

cici37 replied:
So the behavior should be: the newly added field will be ignored and hidden from user if the feature gate is disabled. And ignore the existing old data if you see while updating with feature gate disabled. And add the detailed doc on this new field(like what it did, which feature gate it's controlled, etc)

alexzielenski replied:
Added validation and godoc with feature gate info PTAL

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -50,7 +50,7 @@ type CompilationResult struct {
 	// If true, the compiled expression contains a reference to the identifier "oldSelf", and its corresponding rule
 	// is implicitly a transition rule.
-	TransitionRule bool
+	UsesOldSelf bool
```
Why have to change the name? We have transitionRule documented and used, the name changing might cause confusion. At least state it clearly in comment.

alexzielenski replied:
I changed the name because this name is more explicit. Formerly all rules using oldSelf would be labelled as a "transition rule". Are we still calling the ratcheting rules where oldSelf can be nil a transition rule?

cici37 replied:
I was trying to say we should not break the existing feature behavior. Like set hasOldSelf to be always true might potentially break the existing feature behavior. I don't have a strong preference on names but since we document the transition rule feature everywhere, at least we should document in comment if the name changed.

alexzielenski replied:
Adjusted comment a bit to clearly say that this field `UsesOldSelf` true when the rule contains a reference to `oldSelf`. Is there something more to do for this item?

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -536,11 +542,11 @@ type validationActivation struct {
 	va := &validationActivation{
-		self: UnstructuredToVal(obj, sts),
+		self:       UnstructuredToVal(obj, sts),
+		hasOldSelf: true,
```
hasOldSelf always be true even oldObj == nil? Which means the transaction rule will always be turned on. I am not sure if that makes sense for user to be able to use oldSelf without having an existing old object?

alexzielenski replied:
> I am not sure if that makes sense for user to be able to use oldSelf without having an existing old object?

That is the very intention of this feature. Rules that have `optionalOldSelf` are evaluated with `null` `oldSelf`. If `optionalOldSelf` is false (the default, as it is today for transition rules), then the rule is never evaluated.

alexzielenski replied:
> set hasOldSelf to be always true might potentially break the existing feature behavior

It is writen so that the existing feature remains the same. If expressions dont mention `oldSelf` then they can never observe `hasOldSelf`. A transition rule will also work the same, since it is not evaluated when `oldSelf` is nil in the first place.

jpbetz reviewed (approved):
/approve
Thanks! The approach and high level implementation look reasonable to me.
Leaning on reviews from @cici37 and @apelisse for implementation details.

alexzielenski commented:
added dropdisabledfields and test

alexzielenski commented:
/label api-review
/assign @liggitt

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1096,6 +1102,12 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 
 			}
+
+			if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CRDValidationRatcheting) {
```
Yeah, the field should now be getting dropped so this check is not necessary

alexzielenski committed:
refactor: rename TransitionRule to UsesOldSelf

not all rules that use OldSelf are transition rules, and this flag was used to check for oldSelf usage anyway, not specifically whether the rule was a transition rule

alexzielenski committed:
ratcheting-cel: add optionalOldSelf field

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1142,7 +1154,7 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 							}
 						}
-						if cr.TransitionRule {
+						if allowsNilOldSelf := schema.XValidations[i].OptionalOldSelf; cr.UsesOldSelf && (allowsNilOldSelf == nil || !*allowsNilOldSelf) {
```
agreed. removed validation changes

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
+	// If `oldSelf` is not used in `rule`, this field is ignored.
+	// +featureGate=CRDValidationRatcheting
+	// +optional
+	OptionalOldSelf *bool `json:"optionalOldSelf,omitempty"`
```
whoops. added

cici37 commented:
/retest

alexzielenski commented:
/retest

MaryamTavakkoli commented:
Hello!
Bug triage for the 1.29 release cycle is here!
The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 1 week from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.

As the PR is tagged for 1.29, is it still planned for that release?

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy_test.go:
```diff
@@ -655,21 +659,509 @@ func TestDropDisabledFields(t *testing.T) {
+			},
+		},
+		{
+			name:               "Ratcheting, For update, FG disabled, oldCRD OptionalOldSelf in use, don't drop OptionalOldSelfs",
```
Do we have a test for `Ratcheting, For update, FG disabled, oldCRD OptionalOldSelf in use, but the value is different than the new OptionalOldSelf, don't drop OptionalOldSelfs`?

alexzielenski replied:
I had considered the cases which introduiced a new usage which previously did not exist to excercise that case. To be extra explicit, added a new explicit test for that.

alexzielenski commented:
> As the PR is tagged for 1.29, is it still planned for that release?

yes

cici37 commented:
/lgtm

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
+	// If `oldSelf` is not used in `rule`, this field is ignored.
+	// +featureGate=CRDValidationRatcheting
+	// +optional
+	OptionalOldSelf *bool `json:"optionalOldSelf,omitempty" protobuf:"bytes,6,opt,name=optionalOldSelf"`
```
I like the approach, and the polarity (absent / false keeps behavior as it is today, present+true opts into evaluating even when oldSelf is null).

It wouldn't be an API review without name bikeshedding :-D

stating what we're trying to communicate a few different ways:
* oldSelf is optional
* allow oldSelf to be null
* evaluate [even] when oldSelf is null
* evaluate [even] with null oldSelf

reflections on those:
* "allow" is overloaded with policy/authz; "evaluate" seems clearer to me
* I like including null in the name somehow if that's what their expression would need to check for explicitly (is it? see questions above about type checking and null handling); it seems clearer to me than "optional"
* avoiding a name that leads double-negatives would be ideal (`optionalOldSelf: false` feels like a double negative to me for some reason)

Something like `evaluateWithNullOldSelf: true` seems clearer than `optionalOldSelf: true`, but maybe I've been staring at this for too long.

There's also our [historical avoidance of booleans](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types) in favor of enum-like policies

> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).

Not sure that applies here, but worth at least considering (e.g. `oldSelfEvaluationPolicy: "WhenPresent" | "Always"`)

alexzielenski replied:
I could get behind an enum like `oldSelfEvaluationPolicy` even if we never end up using more cases...

Spitballing some alternative names and cases none of which i totally like hoping we can reach a consensus soon:

1. oldSelfEvaluationPolicy: "WhenPresent" (default) | "Always" (from Jordans comment above)
2. oldSelfBehavior: "Transitions" (default) | Always
3. transitionRulePolicy: "Existing" |  "Unconditional"

/cc @jpbetz @cici37 

for your takes

cici37 replied:
I agree that `optionalOldSelf` sounds less clear especially with false value comparing with the suggestion raised. 
I am slightly leaning towards `oldSelfEvaluationPolicy: "WhenPresent" | "Always"`(transitions is a term user has to look up to know exactly what it refers to here, and we already use `Always` other places comparing with `Unconditional` so keep it consistent would be great) so that user knows exactly what they signed up for and plus leave room for future possibility.

liggitt replied:
with the approach of making `oldSelf` an actual optional value, and getting typechecking to require the rule to treat it that way, I think `optionalOldSelf bool` is more reasonable

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
+
+	// optionalOldSelf is used to indicate a transition rule is always evaluated.
+	// If enabled:
+	// - During a create, `oldSelf` evaluates to `null`
+	// - During an update, `oldSelf` evaluates to `null` when the old object is
```
Does it break CEL type-checking if we're declaring `oldSelf` is of the type of the containing property here

https://github.com/kubernetes/kubernetes/blob/6b8b1881491e26af2370f08d951b7a816d077665/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L161-L163

but then `oldSelf` can sometimes be null? Do we have good tests for the type behavior of rules which use `oldSelf` and allow it to be null? We should do this on properties of every type we support (int, float, string, bool, slice, object) and evaluate the expression with oldSelf present and absent.

Will a check like `oldSelf == null ? ... : ...` work for both null and non-null `oldSelf` values?

Is there a canonical way rules should check for existence of `oldSelf`?
* `oldSelf == null`?
* `has(oldSelf)`?
* `type(oldSelf) == null_type`?

Can we document that for them (and test all the ways we think it *should* work)?

alexzielenski replied:
The only portable way to check this condition that works on all types of `oldSelf` is `type(oldSelf) == null_type`. 

- `has(oldSelf)` is a syntax error since `has` is only implemented for map properties, not top level variables. 
- `oldSelf == null` works as expected for a few types like `object` but is a type-checking error when `oldSelf`/`self` is defined as `string` or `integer`, etc. in the OpenAPI schema.
-  `type(oldSelf) == null_type` is universally well-typed, and works as expected for both nullable and non-nullable types.

I've added to the field doc to suggest users to use the `type` form in their CEL expressions. I've also added a test that it works for all OpenAPI schema types.

liggitt replied:
Ok, so for null checks inside the rule, seems like we have two paths:
* types where self is nullable (object? slice? map?) can do `oldSelf == null` or `type(oldSelf) == null_type`, which is nice
* types where self is not nullable (primitive types?) have to do the `type(oldSelf) == null_type` which is ugly, but at least type checking will bark at them if they do `oldSelf == null` at validation time

I do think it's **very** likely people will set `optionalOldSelf: true` and then not guard `oldSelf` appropriately in their rule, exposing themselves to evaluation errors when `oldSelf` is `null`, since all the validation-time type-checking is done with `oldSelf` being of the same type as `self`. That's the risk I'm trying to think if there's any way we can mitigate. I wish we had a known good value for `self` so we could test if their rule errors when evaluated with `self=$good` and `oldSelf=null`.

liggitt replied:
> That's the risk I'm trying to think if there's any way we can mitigate.

this feels cheap and dirty, but... what if we required the rule to contain one of the following (ignoring whitespace) if they opt into evaluation with null oldSelf, just as a sanity check that they are aware of the possibility of null oldSelf in their rule?
* `oldSelf == null`
* `oldSelf != null`
* `type(oldSelf) == null_type`
* `type(oldSelf) != null_type`

liggitt replied:
once that's updated, be sure to update the API doc as well (for this field and for the `rule` field)

alexzielenski replied:
Updated. And updated API docs and added note to rule field

jpbetz replied:
Just caught up on the latest code. The use of CEL Optionals here is really nice.

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
 	// +optional
 	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+
+	// optionalOldSelf is used to indicate a transition rule is always evaluated.
```
Updated to 

```
	// optionalOldSelf is used to opt a transition rule into evaluation
	// even when the object is first created, or if the old object is
	// missing the value.
```

lmk what you think

liggitt replied:
sounds good

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
+	// - During a create, `oldSelf` evaluates to `null`
+	// - During an update, `oldSelf` evaluates to `null` when the old object is
+	//	 missing the field or if the field is uncorrelatable.
+	// If `oldSelf` is not used in `rule`, this field is ignored.
```
moving this discussion to https://github.com/kubernetes/kubernetes/pull/121034#discussion_r1379521148

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -624,11 +630,11 @@ type validationActivation struct {
 
 func validationActivationWithOldSelf(sts *schema.Structural, obj, oldObj interface{}) interpreter.Activation {
```
Prior to this feature if a rule referred to `oldSelf` and `oldObj` was `nil` then it is not evaluated, so the activation is never used:

https://github.com/kubernetes/kubernetes/blob/3240e2b28ed3b509580455f1bfd01a2ddfcc2c50/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L303-L307 

The new field pretty much only changes the above condition to allow the `optionalOldSelf` field to bypass this `continue`. Setting `hasOldSelf` to `true` prevents CEL from treating `oldSelf` as underfined in this new case where it is being evaluated.

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1182,7 +1182,7 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 							}
 						}
-						if cr.TransitionRule {
+						if cr.UsesOldSelf && (schema.XValidations[i].OptionalOldSelf == nil || !*schema.XValidations[i].OptionalOldSelf) {
```
Agreed, that will help us ensure we are capturing their actual intent. I added a check that the field is unset when `oldSelf` is unused, and validation tests.

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1182,7 +1182,7 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 							}
 						}
-						if cr.TransitionRule {
+						if cr.UsesOldSelf && (schema.XValidations[i].OptionalOldSelf == nil || !*schema.XValidations[i].OptionalOldSelf) {
```
This is a holdout from when we had used an `enum` to express this, so the "CreateUpdate" rule would be allowed to be placed anywhere. 

In the context of `optionalOldSelf` though it does make more sense to me to forbid that. I've changed the validation and added tests that uncorrelatable fields throw errors even with `optionalOldSelf`.

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -223,3 +225,64 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+	}
+}
+
+// dropXValidationsField drops field XValidations from CRD schema
```
Yeah, `SchemaHas` is read-only. Another such place I can think of that handles its own recursive mutation (but returns a minimal copy) is in [CEL schema resolver](https://github.com/kubernetes/kubernetes/blob/e4776e0f85b8aceb8f1da7a87822b0d086045a8a/staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver/refs.go#L82-L102). `kubectl-validate` also has own hand-rolled mutator pattern.

I do agree its probably time to add a helper to `kube-openapi` for this if one doesn't already exist.

alexzielenski replied:
found

https://github.com/kubernetes/kube-openapi/blob/8033fb28565aa6548d65e973756081a475661196/pkg/schemamutation/walker.go#L23-L36

will refactor to use this

liggitt replied:
that's dealing in kube-openapi types, and this mutator has to deal in API server CRD types, I don't think that will work

alexzielenski replied:
Yeah my mistake...I was recalling incorrect context when I read this comment thinking it was already spec.Schema 😅

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -223,3 +225,64 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+			schema.Items.JSONSchemas[i] = jsonSchema
+		}
+	}
+	for def, jsonSchemaPropsOrStringArray := range schema.Dependencies {
+		dropOptionalOldSelfField(jsonSchemaPropsOrStringArray.Schema)
+		schema.Dependencies[def] = jsonSchemaPropsOrStringArray
+	}
```
definitions and dependencies aren't supported in CRDs, right? can drop this stanza

alexzielenski replied:
Thats right. It was there for consistency with the original dropXValidations function that was removed recently during a rebase. 

Removed.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
 	// +optional
 	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+
+	// optionalOldSelf is used to indicate a transition rule is always evaluated.
```
ha! How come we missed it before 😅 . I'll raise a PR later to fix it. ( We do have transition rule explained properly in doc: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules)

alexzielenski commented:
/retest

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -161,8 +166,42 @@ func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclTy
+			EnvOptions: []cel.EnvOption{
+				cel.Variable(OldScopedVarName, types.NewOptionalType(scopedType.CelType())),
+			},
+			DeclTypes: []*apiservercel.DeclType{
+				scopedType,
+			},
```
Curious if this was strictly needed since the decl is already added to the environment in the prior version options?  If we don't need to declare it again maybe omit it?  (if we do need to declare it again, I'll want to dig in to figure out why..)

alexzielenski replied:
Was able to remove it. It needed to be added in an earlier implementation when I had called extend multiple times like the following pseudocode:

```go
selfEnvSet := baseEnv.Extend({
     // self definition
    DeclTypes: [scopedType]
})

oldSelfEnvSet := selfEnvSet.Extend({
  // oldself definition
})


optionalOldSelfEnvSet := selfEnvSet.Extend({
    // optionalOldSelf definition
})
```

This caused field resolution errors that weren't resolved until we either added the DeclType to each call to Extend, or made it the form it is currently.

@jpbetz is it expected that the definition wouldn't be available in successive calls to `Extend`?

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -65,9 +67,10 @@ type Validator struct {
-	// oldSelf) to CEL values.
-	celActivationFactory func(sts *schema.Structural, obj, oldObj interface{}) interpreter.Activation
+	// celActivationFactory produces a Activations, which resolve identifiers
+	// (e.g. self and oldSelf) to CEL values. One activation must be produced
+	// for
```
Last sentence is incomplete?

alexzielenski replied:
whoops. added the rest:
```

	// celActivationFactory produces a Activations, which resolve identifiers
	// (e.g. self and oldSelf) to CEL values. One activation must be produced
	// for each of the cases when oldSelf is optional and non-optional.
```

jpbetz reviewed (commented):
Two minor comments, then the implementation LGTM.   I glanced at the API changes as well and don't have any concerns, but will defer to API review for that.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
copy doc updates to v1 as well

alexzielenski replied:
should be there now

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:
there should be type checking unit tests on the optional oldSelf to make sure we catch type stuff:

```diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go
index 18df5a42475..a774e818b9d 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go
@@ -26,13 +26,18 @@ import (
 	"github.com/google/cel-go/common/types"
 	"github.com/google/cel-go/common/types/ref"
 
+	"k8s.io/utils/pointer"
+
 	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
 	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model"
+	apiextensionsfeatures "k8s.io/apiextensions-apiserver/pkg/features"
 	"k8s.io/apimachinery/pkg/util/version"
 	celconfig "k8s.io/apiserver/pkg/apis/cel"
 	"k8s.io/apiserver/pkg/cel"
 	"k8s.io/apiserver/pkg/cel/environment"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
 )
 
 const (
@@ -151,6 +156,8 @@ func (v transitionRuleMatcher) String() string {
 }
 
 func TestCelCompilation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, apiextensionsfeatures.CRDValidationRatcheting, true)()
+
 	cases := []struct {
 		name            string
 		input           schema.Structural
@@ -625,6 +632,92 @@ func TestCelCompilation(t *testing.T) {
 				matchesAll(noError(), transitionRule(true)),
 			},
 		},
+		{
+			name: "optional primitive transition rule type checking",
+			input: schema.Structural{
+				Generic: schema.Generic{
+					Type: "integer",
+				},
+				Extensions: schema.Extensions{
+					XValidations: apiextensions.ValidationRules{
+						{Rule: "self >= oldSelf.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self >= oldSelf.orValue(1)", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "oldSelf.hasValue() ? self >= oldSelf.value() : true", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self >= oldSelf", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self >= oldSelf.orValue('')", OptionalOldSelf: pointer.Bool(true)},
+					},
+				},
+			},
+			expectedResults: []validationMatcher{
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(invalidError("optional")),
+				matchesAll(invalidError("orValue")),
+			},
+		},
+		{
+			name: "optional complex transition rule type checking",
+			input: schema.Structural{
+				Generic: schema.Generic{
+					Type: "object",
+				},
+				Properties: map[string]schema.Structural{
+					"i": schema.Structural{Generic: schema.Generic{Type: "integer"}},
+					"b": schema.Structural{Generic: schema.Generic{Type: "boolean"}},
+					"s": schema.Structural{Generic: schema.Generic{Type: "string"}},
+					"a": schema.Structural{
+						Generic: schema.Generic{Type: "array"},
+						Items:   &schema.Structural{Generic: schema.Generic{Type: "integer"}},
+					},
+					"o": schema.Structural{
+						Generic: schema.Generic{Type: "object"},
+						Properties: map[string]schema.Structural{
+							"i": schema.Structural{Generic: schema.Generic{Type: "integer"}},
+							"b": schema.Structural{Generic: schema.Generic{Type: "boolean"}},
+							"s": schema.Structural{Generic: schema.Generic{Type: "string"}},
+							"a": schema.Structural{
+								Generic: schema.Generic{Type: "array"},
+								Items:   &schema.Structural{Generic: schema.Generic{Type: "integer"}},
+							},
+							"o": schema.Structural{
+								Generic: schema.Generic{Type: "object"},
+							},
+						},
+					},
+				},
+				Extensions: schema.Extensions{
+					XValidations: apiextensions.ValidationRules{
+						{Rule: "self.i >= oldSelf.i.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.s == oldSelf.s.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.b == oldSelf.b.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o == oldSelf.o.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.i >= oldSelf.o.i.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.s == oldSelf.o.s.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.b == oldSelf.o.b.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.o == oldSelf.o.o.value()", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.i >= oldSelf.o.i.orValue(1)", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "oldSelf.hasValue() ? self.o.i >= oldSelf.o.i.value() : true", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.i >= oldSelf.o.i", OptionalOldSelf: pointer.Bool(true)},
+						{Rule: "self.o.i >= oldSelf.o.s.orValue(0)", OptionalOldSelf: pointer.Bool(true)},
+					},
+				},
+			},
+			expectedResults: []validationMatcher{
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(noError(), transitionRule(true)),
+				matchesAll(invalidError("optional")),
+				matchesAll(invalidError("orValue")),
+			},
+		},
 		{
 			name: "whitespace-only rule",
 			input: schema.Structural{
diff --git a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go
index 236564f7d36..63119ebaa22 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go
@@ -3667,6 +3667,18 @@ func TestOptionalOldSelf(t *testing.T) {
 				"foo": stringType,
 			}), "self.foo == 'valid' || (oldSelf.hasValue() && oldSelf.value().foo != 'valid')"),
 		},
+		{
+			name: "allow invalid new value if old value is also invalid with chained optionals",
+			obj: map[string]interface{}{
+				"foo": "invalid again",
+			},
+			oldObj: map[string]interface{}{
+				"foo": "invalid",
+			},
+			schema: withRulePtr(objectTypePtr(map[string]schema.Structural{
+				"foo": stringType,
+			}), "self.foo == 'valid' || oldSelf.foo.orValue('') != 'valid'"),
+		},
 		{
 			name: "block invalid new value if old value is valid",
 			obj: map[string]interface{}{
```

alexzielenski replied:
Added

liggitt commented:
looks like some TestValidateCustomResourceDefinitionValidation unit test failures as well

liggitt commented:
I know we validate a `default` specified in openapi against any CEL rules provided:

https://github.com/kubernetes/kubernetes/blob/263ab254ef6a847fdd3dc69250a2ac0b12c42eb3/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go#L76-L126

It looks like we're currently passing `s.Default.Object` as both the old and new object, to run as many validation rules as possible.

If the ratcheting validation feature is enabled, and it passed that first validation check, should we also validate with `s.Default.Object` as the new object and `nil` as the old object, to make sure defaults pass any ratcheting rules that behave differently on create?

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -622,21 +632,44 @@ type validationActivation struct {
 	if oldObj != nil {
 		va.oldSelf = UnstructuredToVal(oldObj, sts) // +k8s:verify-mutation:reason=clone
-		va.hasOldSelf = true                        // +k8s:verify-mutation:reason=clone
+		va.hasOldSelf = true
```
I think it is worth it as a sanity check to old oldSelf undefined in contexts where it would be unexpectedly nil. Since we are using a separate activation now for optionalOldSelf.

liggitt replied:
ok

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -246,6 +254,24 @@ type ValidationRule struct {
+	// missing the value.
+	//
+	// When enabled, if the old object is missing the value or is being
+	// created for the first time, then `oldSelf` will be a CEL optional.
```
Rephrased:

```
	// When enabled `oldSelf` will be a CEL optional whose value will be
	// `None` if there is no old value, or when the object is initially created.
	//
	// You may check for presence of oldSelf using `oldSelf.hasValue()` and
	// unwrap it after checking using `oldSelf.value()`. Check the CEL
	// documentation for Optional types for more information:
	// https://pkg.go.dev/github.com/google/cel-go/cel#OptionalTypes
	//
```

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -210,6 +210,14 @@ type ValidationRule struct {
 	//     non-intersecting keys are appended, retaining their partial order.
+	//
+	// If `rule` makes use of the `oldSelf` variable it is implicitly a
+	// `transition rule`.
```
Added

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -622,21 +632,44 @@ type validationActivation struct {
 	}
+	optionalVA := &validationActivation{
+		self:       va.self,
+		hasOldSelf: true,
```
Added

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -622,21 +632,44 @@ type validationActivation struct {
 }
 
-func validationActivationWithOldSelf(sts *schema.Structural, obj, oldObj interface{}) interpreter.Activation {
+func validationActivationWithOldSelf(sts *schema.Structural, obj, oldObj interface{}) (interpreter.Activation, interpreter.Activation) {
```
added names to the return values

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -622,21 +632,44 @@ type validationActivation struct {
+	return va, optionalVA
+}
+
+func validationActivationWithOptionalOldSelf(sts *schema.Structural, obj, oldObj interface{}) interpreter.Activation {
```
removed

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -300,11 +303,18 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+		if compiled.UsesOldSelf && oldObj == nil && !allowsNilOldSelf {
 			// transition rules are evaluated only if there is a comparable existing value
 			continue
+		} else if compiled.UsesOldSelf && allowsNilOldSelf {
```
restructured as suggested

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -161,8 +166,42 @@ func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclTy
 			},
+			DeclTypes: []*apiservercel.DeclType{
+				scopedType,
+			},
```
Moving to duplicate: https://github.com/kubernetes/kubernetes/pull/121034#discussion_r1382139900

alexzielenski committed:
validate defaults across an update from nil to ensure create ratcheting rules work

alexzielenski commented:
> If the ratcheting validation feature is enabled, and it passed that first validation check, should we also validate with s.Default.Object as the new object and nil as the old object, to make sure defaults pass any ratcheting rules that behave differently on create?

I think that'd be prudent. Ive added it and a test for it

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -160,8 +166,42 @@ func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclTy
+			},
+		},
+		environment.VersionedOptions{
+			IntroducedVersion: version.MajorMinor(1, 24),
```
Wondering if/how versioning/rollout is affected by this variable decl cc @jpbetz

liggitt replied:
use of this env is gated by an actual API field with an actual feature gate attached, so I don't think this is really relevant. I think we should match the existing `oldSelf` initial version here

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -136,15 +136,20 @@ func Compile(s *schema.Structural, declType *apiservercel.DeclType, perCallLimit
 }
 
-func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclType) (*environment.EnvSet, error) {
+func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclType) (*environment.EnvSet, *environment.EnvSet, error) {
```
doc or name the return vars

alexzielenski replied:
named return vars

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -300,11 +304,27 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+		if compiled.UsesOldSelf && oldObj == nil {
 			// transition rules are evaluated only if there is a comparable existing value
-			continue
+			// But if the rule uses optoinal oldSelf and gate is enabled we allow
```
```suggestion
			// But if the rule uses optional oldSelf and gate is enabled we allow
```

alexzielenski replied:
Fixed

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -300,11 +304,27 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+			// the rule to be evaluated
+			if !utilfeature.DefaultFeatureGate.Enabled(features.CRDValidationRatcheting) {
+				continue
+			} else if !optionalOldSelfRule {
```
```suggestion
			}
			if !optionalOldSelfRule {
```

alexzielenski replied:
Changed

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go:
```diff
@@ -96,6 +98,11 @@ func validate(ctx context.Context, pth *field.Path, s *structuralschema.Structur
 					return allErrs, nil, remainingCost
 				}
+
+				if len(celErrs) == 0 && utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CRDValidationRatcheting) {
```
needs a comment explaining why we're calling this again

also, I think we need to capture the cost of both types of runs and take whichever leaves less remaining cost

```go
celErrs, rmCost := celValidator.Validate(ctx, pth.Child("default"), s, s.Default.Object, s.Default.Object, remainingCost)
allErrs = append(allErrs, celErrs...)

if len(celErrs) == 0 && utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CRDValidationRatcheting) {
  celErrs, rmCostWithoutOldObject := celValidator.Validate(ctx, pth.Child("default"), s, s.Default.Object, nil, remainingCost)
  allErrs = append(allErrs, celErrs...)
  if rmCostWithoutOldObject < rmCost {
    rmCost = rmCostWithoutOldObject
  }
}

remainingCost = rmCost
if remainingCost < 0 {
  return allErrs, nil, remainingCost
}
```

liggitt replied:
same comment applies below

alexzielenski replied:
Applied feedback as suggested and added comments

liggitt commented:
/lgtm
/approve

alexzielenski committed:
ratcheting-cel: use Optional[T] for oldSelf when optionalOldSelf is true

apelisse commented:
Blind reapply
/lgtm


On Fri, Nov 3, 2023 at 3:45 PM Kubernetes Prow Robot <
***@***.***> wrote:

> New changes are detected. LGTM label has been removed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/121034#issuecomment-1793208504>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOXDLEIOC5A7Y5LHCJRWC3YCVXZXAVCNFSM6AAAAAA5WG6SX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJTGIYDQNJQGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
